README.md
# Project Importer

The Project Importer handles importing project plans from other tools into  Web2project.  At present, the following tools are supported for import:

* Microsoft Project 2003 and earlier;
* Microsoft Project 2007 - when saved as a Project 2003 XML file;
* WBS Gantt Chart Pro v4.4 - other versions may be supported, this specific version has been validated;

## Open Issues:

*  By default, the Importer will attempt to create users for any resources (aka people) it does not recognize.  If the resource name (username) is below the minimum allowed length in Web2project (default: 4), they will not be able to log into the system.  It is recommended that you adjust any usernames to meet this requirement.
*  The Microsoft Project import class still uses the php4 style of XML parsing. It should be updated to use SimpleXML or something similar to speed processing and reduce memory use.

## Updated in 4.2:

*  Added numerous notes on areas to cleanup & simplify in future versions;
*  Applied proper theming to all form fields;
*  Fixed the Contact lookups to work more often;
*  Fixed the Task Assignees to allow existing users (not just newly imported users) to be assigned to Tasks;
*  Fixed the Task Assignees to respect permissions and limit the selection only to Users who can actually view Tasks;
*  Fixed the "View the project here" link;
*  Fixed the Work Breakdown Structure id for the WBS Pro file format;
*  Refactored the flow of the module to the 'proper' structure now specified in the web2project Module Building Guide;
*  Refactored the post-upload step to a clean redirect with the preview;
*  Removed the references to the raw $_FILES;
*  Updated the class structures to standardize look & feel;
*  Updated the deprecated DBQuery to the new w2p_Database_Query class;
*  Updated the project/task processing to be timezone aware;
*  Updated the status messages to use the AppUI status messages instead of echo;

## Updated in 4.1:

*  Fixed a major issue where dependencies were not properly imported or sometimes at all.
*  Added a tweak to the MS Project processor to make sure the project title is retrieved properly.
*  Fixed an issue where the Project Start Date was not set properly.
*  Tweaked the import process so that if there's no Company match, it default's to the current User's Company. This resolves the issue where a Project could accidentally get assigned to no Company and therefore disappear due to permissions.
*  Changed the licensing to be in accordance with the the coming web2project shift.

## Updated in 4.0:

*  Added some hr tags and lined up some input boxes to make the UI cleaner.
*  Updated the module structure (paths and names) to use the core web2project autoloader and simplify the includes code.
*  Refactored much of the Project/Task/Contact creation code into its proper class methods instead of raw DBQuery calls. Updated those calls to use the object validation results. Further moved this code into the base CImporter class to reduce duplication.
*  Applied some major patches from Alain Picard to handle larger and larger file sizes. In most cases, this is a 20-25% improvement, but some imports can get as much as an 80% improvement in processing time.
  * ~95 Tasks process in   ~4 seconds using  12MB at peak;
  * ~450 Tasks process in  ~50 seconds using  15MB at peak;
  * ~1100 Tasks process in ~180 seconds using 380MB at peak;

## Install

1.  Previous installations of this module can simply be removed via the System Admin -> View Modules screen.
1.  To install this module, please follow the standard module installation procedure.
1.  Download the latest version from [Github](https://github.com/web2project/modules-importers) and unzip the file. Take the existing directory (called: importers) and move/upload it into your web2project/modules directory.
1.  Select to System Admin -> View Modules and you should see "Project Importer" near the bottom of the list.
1.  On the "importer" row, select "install".  The screen should refresh and the line should become "Project Importer".  Now select "hidden" and then "disabled" to make it display in your module navigation.
1.  You should be able to order it within the navigation like any other module.

## Usage

1.  Within Microsoft Project, open your project as usual.  In the File Menu, select "Save AS ...".  Under file-type, choose "xml" and save your file.
1.  Within Web2project, select "Project Import" from your module navigation.
1.  Select "Browse", select the file you saved in Step 1, and select "Import Data".
1.  Your screen should now show a summary of the imported project with all the relevant Company, Project, Task, and Assignment information.
1.  If a matching Company was found, the Importer will automatically assign it.  Alternatively, if there was no matching Company, you can choose one or allow the Importer to create a new one.
1.  Input a Project Name.  If the imported name is already in use, the module will note it here.
1.  Assign the usernames who match your resources.  Any users not mapped to current Web2project users will automatically be created in an Inactive state.
1.  Adjust the task assigments as necessary.  By default, the percent allocations from the MS Project file will be used.
1.  Select "import" below.

### Development Status

* v4.1 Update Release: 2010 September
* v4.0 Major  Release: 2010 April
* v3.2 Update Release: 2009 October
* <s>v3.1 Update Release: 2009 September</s>
* v3.0 Major  Release: 2008 September
* *All versions prior to v3.0 are only compatible with dotProject - http://docs.dotproject.net/index.php?title=Project_Importer*
* v2.0 Major  Release: 2007 October (requires PHP5 and dotProject v2.1 rc1 or above)
* v1.5 Update Release: 2007 July (last PHP4-compatible release)
* v1.4 Update Release: 2007 May
* v1.2 Update Release: 2006 October (last dotProject 2.0.x-compatible release)
* v1.1 Update Release: 2006 September
* v1.0 Initial Release: 2005 December

If you find this module particularly useful and would like to express
gratitude or seek additional development, please do not hesitate to contact
CaseySoftware, LLC via webmaster@caseysoftware.com

Microsoft Project is an application owned by Microsoft Corporation.  All Rights Reserved.


CHANGELOG.md
# Change Log


Updated in 3.2:

*  Fixed a bug in import where it simply wasn't importing Projects unless you had some odd combination of permissions... now it's be simplified to just requiring the Project:Add permission;
*  Converted a bunch of the raw DBQuery calls to calls on the proper objects, this doesn't do anything with the results of the object validation, but that's not formally in Web2project until v1.2 anyway;

Updated in 3.1:

*  Simplified the permissions checking to only check for Project Add privileges;
*  Removed the poorly implemented translation files to take advantage of the reworked translation handler in Web2project v1.1;
*  Simplified the View to use a case statement instead of nested if's.  The code isn't shorter, but it's quite a bit clearer;
*  General clean up to use core Web2project functionality as opposed to using custom code, it isn't complete;

Fixed in 3.0:

*  Updated the Project Importer to support Web2project.  In the process, backwards compatibility with dotProject was lost.  This is a known condition and is not considered an issue or bug, it is a design decision and will not be reversed or reconsidered.
*  Simplified the class structure and encapsulated more of the type-specific details away from the user and the BaseImporter class.
*  Converted some of the raw SQL statements to use the DBQuery class.

Fixed in 2.2:

*  Added a contribution from Wellison da Rocha Pereira to allow for importing of WBS Gant Chart Pro v4.4 files.
*  Added a contribution from Richard Siomporas to allow the User to choose whether or not to import Users from the imported file.
*  Found and resolved an issue where it was possible to assign a project to Company 0 which would prevent all users from accessing the project in any way.

Fixed in 2.1:

*  Fixed an issue with the task_type which was only being set consistently for dynamic tasks.  This would cause some tasks not to display in the Project View.
*  If you attempt to import a project and there is already one with an identical name, the Importer will automatically delete the existing tasks and import the new tasks into the existing project.  A warning will appear to the user during the preview stage.

Fixed in 2.0:

*  Fixed the parent/child relationship error.  In the v1.5 release, a bug was introduced which could cause parent/child relationships to be lost.
*  Fixed the User case mismatch error.  Previously, when the Importer created Users, it would use the same capitalization from the Microsoft Project file's Resources.  Unfortunately, this often created sitatuations wherea a new User named "Keith" was created while "keith" was an existing User.  The Importer now makes all usernames lowercase to match the rest of the system.
*  Improved overall memory management.  Previously, with 64M allocated to PHP, approximately 100 tasks in a single project plan could be imported.  With the new memory management, 64M has proven sufficient to import 450+ tasks.
*  Created a simple class hierarchy.  Previously, only a single preview() and import() could be supported.  By creating a simple class structure, the goal is to more easily support additional filetypes in future releases.
*  Adjusted the install() and remove() functions in the setup procedure to automatically attempt to install the translation files (English and German currently) to their proper locations within dotProject.  This ensures that all language changes via the Translation Management are reflected.

Fixed in 1.5:

*  Fixed the duration rounding error.  Previously tasks with fractional hour durations were being rounded down to the nearest hour.  Now the actual value is preserved.
*  Fixed the character cutoff issue.  Due to how PHP4 handles XML parsing and cdata blocks, special characters were triggering the parser to treat the single cdata node as multiple node therefore causing only the last of the text segments to be imported.
*  Various raw SQL updates were replaced with their DBQuery equivalents.

Fixed in 1.4:

*  Tweaked the resource import.  Previously there was no way to completely remove a person.  Changing a user's assignment to zero removes them from the task completely.
*  Fixed the resource import.  Previously, if a task had no one assigned the import process would generate an error.  It now checks the array size before attempting to iterate over it.
*  Added the ability to chose a Project Status during import.
*  Added the ability to chose a Project Owner during import, defaults to the user doing the import.
*  Modified the preview screen to warn the user if a username is below the minimum length set in their dotProject configuration.
*  Modified the preview screen to warn the user about permissions if the module is about to create a new Company.
*  Modified the import functionality to use the properly scrubbed parameters, etc via the core dPgetParam functions.
*  Cleaned up the code to better match the coding standards in use by dotProject (PEAR).
*  Added the security sentinal from core dotProject to prevent unauthorized access via the filesystem.  This breaks compatibility with all dotProject releases prior to 2.1-rc1.
*  Renamed the module from the "Microsoft Project Importer" to simply "Project Importer".  This is to avoid any potential legal issues and acknowledging the planned direction of the module.

Fixed in 1.3:

*  All user assignments are now calculated from the values in the Microsoft Project file itself.  This insight was shared by the highly esteemed Tim Millhouse of Digilore, Inc and implemented by D. Keith Casey, Jr. of CaseySoftware, LLC.

imports/WBSImporter.class.php
<?php
if (!defined('W2P_BASE_DIR')){
	die('You should not access this file directly.');
}

/* Added by Wellison da Rocha Pereira, credited in license.txt */
//TODO: there needs to be a naming convention here for the autoloader
class WBSImporter extends CImporter {
	protected $depthArray = array();
	protected $lastLevel = 0;
	protected $maxLevel = 0;

	/*
	 * The $fields array here is actually the $_POST. I switched it to fields
	 *   so we could test it more easily.
	 *   dkc - 15 April 2011
	 */
    public function import(w2p_Core_CAppUI $AppUI, array $fields) {

		parent::_import($this->AppUI, $fields);

        $q = new w2p_Database_Query();
        // Users Setup
        if (isset($fields['users']) && is_array($fields['users']) && $fields['nouserimport'] != "true") {
            foreach($fields['users'] as $ruid => $r) {
                $q->clear();

                if (!empty($r['user_username'])) {
                    $contact_id = (int) $this->_processContact($this->AppUI, $r['user_username'], $this->company_id);
                    if ($contact_id) {
//TODO:  Replace with the regular create users functionality
						$q->addInsert('user_username', $r['user_username']);
						$q->addInsert('user_contact', $contact_id);
						$q->addTable('users');
						$q->exec();
						$insert_id = db_insert_id();

						$r['user_id'] = $insert_id;
                    } else {
//TODO:  This error message doesn't make it through..
						$this->AppUI->setMsg($result, UI_MSG_ERROR);
					}
                } else {
                    $r['user_id'] = $r['user_userselect'];
                }
                if (!empty($r['user_id'])) {
                    $resources[$ruid] = $r;
                }
            }
        }

        // Tasks Setup
        foreach ($fields['tasks'] as $k => $task) {
            $result = $this->_processTask($this->AppUI, $this->project_id, $task);
            if (is_array($result)) {
                $this->AppUI->setMsg($result, UI_MSG_ERROR);
                $this->AppUI->redirect('m=importers');
            }
            $task_id = $result;

            // Task Parenthood
            $outline[$task['OUTLINENUMBER']] = $task_id;
            $q->clear();
//TODO:  Replace with the regular task parent handling
            if (!strpos($task['OUTLINENUMBER'], '.')) {
                $q->addUpdate('task_parent', $task_id);
                $q->addWhere('task_id = ' . $task_id);
                $q->addTable('tasks');
            } else {
                $parent_string = substr($task['OUTLINENUMBER'], 0, strrpos($task['OUTLINENUMBER'], '.'));
                $parent_outline = isset($outline[$parent_string]) ? $outline[$parent_string] : $task_id;
                $q->addUpdate('task_parent', $parent_outline);
                $q->addWhere('task_id = ' . $task_id);
                $q->addTable('tasks');
            }
            $q->exec();

            $task['task_id'] = $task_id;
			$tasks[$task['UID']] = $task;
//TODO:  Replace with the regular task assignment handling
            // Resources (Workers)
            if (count($task['resources']) > 0) {
				$resourceArray = array();
//TODO: figure out how to assign to existing users
				foreach($task['resources'] as $uk => $user) {
                    $alloc = $task['resources_alloc'][$uk];
					$q->clear();
                    if ($alloc > 0 && $resources[$user]['user_id'] > 0) {
						$user_id = $resources[$user]['user_id'];
						if (!in_array($user_id, $resourceArray)) {
							$q->addInsert('user_id', $user_id);
							$q->addInsert('task_id', $task_id);
							$q->addInsert('perc_assignment', $alloc);
							$q->addTable('user_tasks');
							$q->exec();
						}
						$resourceArray[] = $resources[$user]['user_id'];
                    }
					if ((int) $user) {
						$q->addInsert('user_id', $user);
						$q->addInsert('task_id', $task_id);
						$q->addInsert('perc_assignment', $alloc);
						$q->addTable('user_tasks');
						$q->exec();
					}
                }
            }

            // Task Dependencies
            if (is_array($task['dependencies'])) {
                foreach($task['dependencies'] as $task_uid) {
                    if ($task_uid > 0 && $tasks[$task_uid]['task_id'] > 0) {
                        $q->clear();
                        $q->addInsert('dependencies_task_id', $task_id);
                        $q->addInsert('dependencies_req_task_id', $tasks[$task_uid]['task_id']);
                        $q->addTable('task_dependencies');
                        $q->exec();
                    }
                }
            }
        }
        $this->_deDynamicLeafNodes($this->project_id);
        addHistory('projects', $this->project_id, 'add', $projectName, $this->project_id);
        return true;
    }

    public function view(w2p_Core_CAppUI $AppUI) {
        /*
		 * The whole view is based on the SimpleXML. It was a easier decision 
		 *   to get atributes from the wbs files and it actually worked! I 
		 *   tried XMLReader but i've failed completely.
		 *
		 */
        $perms = $this->AppUI->acl();

        $output = '';
        $data = $this->scrubbedData;

        $reader = simplexml_load_string($data);

        $project_name = $reader->proj->summary['Title'];
        if (empty($project_name)) {
            $project_name=$this->proName;
		}

        $output .= '
			<table width="100%">
			<tr>
			<td align="right">' . $this->AppUI->_('Company Name') . ':</td>';

		$projectClass = new CProject();
		$output .= $this->_createCompanySelection($this->AppUI, $tree['COMPANY']);
		$output .= $this->_createProjectSelection($this->AppUI, $project_name);

		$users = $perms->getPermittedUsers('projects');
		$output .= '<tr><td align="right">' . $this->AppUI->_('Project Owner') . ':</td><td>';
		$output .= arraySelect( $users, 'project_owner', 'size="1" style="width:200px;" class="text"', $this->AppUI->user_id );
		$output .= '<td/></tr>';

		$pstatus =  w2PgetSysVal( 'ProjectStatus' );
		$output .= '<tr><td align="right">' . $this->AppUI->_('Project Status') . ':</td><td>';
		$output .= arraySelect( $pstatus, 'project_status', 'size="1" class="text"', $row->project_status, true );
		$output .= '<td/></tr>';

		$startDate = $this->_formatDate($this->AppUI, $reader->proj->summary['Start']);
		$endDate = $this->_formatDate($this->AppUI, $reader->proj->summary['Finish']);

		$output .= '
            <tr>
                <td align="right">' . $this->AppUI->_('Start Date') . ':</td>
                <td>
					<input type="hidden" name="project_start_date" value="'.$startDate.'" class="text" />
					<input type="text" name="start_date" value="'.$reader->proj->summary['Start'].'" class="text" />
				</td>
            </tr>
            <tr>
                <td align="right">' . $this->AppUI->_('End Date') . ':</td>
                <td>
					<input type="hidden" name="project_end_date" value="'.$endDate.'" class="text" />
					<input type="text" name="end_date" value="'.$reader->proj->summary['Finish'].'" class="text" />
				</td>
            </tr><!--
            <tr>
                <td align="right">' . $this->AppUI->_('Do Not Import Users') . ':</td>
                <td><input type="checkbox" name="nouserimport" value="true" onclick="ToggleUserFields()" /></td>
            </tr>
            <tr>
                <td colspan="2">' . $this->AppUI->_('Users') . ':</td>
            </tr>
            <tr>
                <td colspan="2"><div name="userRelated"><br /><em>'.$this->AppUI->_('userinfo').'</em></td>
			</tr>-->
			<tr>
				<td colspan="2"><table>';

        $percent = array(0 => '0', 5 => '5', 10 => '10', 15 => '15', 20 => '20', 25 => '25', 30 => '30', 35 => '35', 40 => '40', 45 => '45', 50 => '50', 55 => '55', 60 => '60', 65 => '65', 70 => '70', 75 => '75', 80 => '80', 85 => '85', 90 => '90', 95 => '95', 100 => '100');

		// Users (Resources)
		$workers = $perms->getPermittedUsers('tasks');
        $resources = array(0 => '');

		$q = new w2p_Database_Query();
        if ($this->user_control) {		//check the existence of resources before trying to import
            $trabalhadores=$reader->proj->resources->children();
            foreach($trabalhadores as $r) {
                $q->clear();
                $q->addQuery('user_id');
                $q->addTable('users');
                $q->leftJoin('contacts', 'c', 'user_contact = contact_id');
                $q->addWhere("user_username LIKE '{$r['name']}' OR CONCAT_WS(contact_first_name, ' ', contact_last_name) = '{$r['name']}'");
                $r['LID'] = $q->loadResult();
                if (!empty($r['name'])) {
                    $output .= '
						<tr>
						<td>' . $this->AppUI->_('User name') . ': </td>
						<td>
						<input type="text" name="users[' . $r['uid'] . '][user_username]" value="' . ucwords(strtolower($r['name'])) . '"' . (empty($r['LID'])?'':' readonly') . ' />
						<input type="hidden" name="users[' . $r['uid'] . '][user_id]" value="' . $r['LID'] . '" />
						(' . $this->AppUI->_('Resource UID').": ".$r['uid'] . ')';

                    if (function_exists('w2PUTF8strlen')) {
                        if (w2PUTF8strlen($r['name']) < w2PgetConfig('username_min_len')) {
                            $output .= ' <em>' . $this->AppUI->_('username_min_len.') . '</em>';
                        }
                    } else {
                        if (strlen($r['name']) < w2PgetConfig('username_min_len')) {
                            $output .= ' <em>' . $this->AppUI->_('username_min_len.') . '</em>';
                        }
                    }
                    $output .= '</td></tr>';
                    $resources[sizeof($resources)] = strtolower($r['name']);
                }
            }
        }
		$resources = arrayMerge($resources, $workers);

        $output .= '
            </table>
            </div></td></tr>';

		// Insert Tasks
        $output .= '
            <tr>
            <td colspan="2">' . $this->AppUI->_('Tasks') . ':</td>
            </tr>
            <tr>
            <td colspan="2">
            <table width="100%" class="tbl" cellspacing="1" cellpadding="2" border="0">
            <tr>
            <th>' . $this->AppUI->_('Name') . '</th>
            <th>' . $this->AppUI->_('Start Date') . '</th>
            <th>' . $this->AppUI->_('End Date') . '</th>
            <th>' . $this->AppUI->_('Assigned Users') . '</th>
            </tr>';

        foreach($reader->proj->tasks->children() as $task) {
            if (trim($task['Name']) != '') {
                $k = $task['ID'];
				$newWBS = $this->montar_wbs($task['OutlineLevel']);
                $note= htmlentities($task['NOTES']);
                $output .= '<tr><td>';
                $output .= '<input type="hidden" name="tasks['.$k.'][UID]" value="' . $k . '" />';
                $output .= '<input type="hidden" name="tasks['.$k.'][OUTLINENUMBER]" value="' . $newWBS . '" />';
                $output .= '<input type="hidden" name="tasks['.$k.'][task_name]" value="' . $task['Name'] . '" />';
                $output .= '<input type="hidden" name="tasks['.$k.'][task_description]" value="' . $note . '" />';

                $priority = 0;
                $output .= '<input type="hidden" name="tasks['.$k.'][task_priority]" value="' . $priority . '" />';
                $output .= '<input type="hidden" name="tasks['.$k.'][task_start_date]" value="' . $task['Start'] . '" />';
                $output .= '<input type="hidden" name="tasks['.$k.'][task_end_date]" value="' . $task['Finish'] . '" />';

                $myDuration = $this->dur($task['Duration']);
				$output .= '<input type="hidden" name="tasks['.$k.'][task_duration]" value="' . $myDuration . '" />';
				$output .= '<input type="hidden" name="tasks['.$k.'][task_duration_type]" value="1" />';

                $percentComplete = isset($task['PercentComplete']) ? $task['PercentComplete'] : 0;
                $output .= '<input type="hidden" name="tasks['.$k.'][task_percent_complete]" value="' . $percentComplete . '" />';
                $output .= '<input type="hidden" name="tasks['.$k.'][task_owner]" value="'.$this->AppUI->user_id.'" />';
                $output .= '<input type="hidden" name="tasks['.$k.'][task_type]" value="0" />';

                $milestone = ($task['Milestone'] == 'yes') ? 1 : 0;
                $output .= '<input type="hidden" name="tasks['.$k.'][task_milestone]" value="' . $milestone . '" />';

                $temp = 0;
                if (!empty($task['UniqueIDPredecessors'])) {
                    $x=strpos($task['UniqueIDPredecessors'],",");
                    foreach ($task['UniqueIDPredecessors'] as $dependency) {
                        $output .= '<input type="hidden" name="tasks['.$k.'][dependencies][]" value="' . $dependency['UniqueIDPredecessors'] . '" />';
                        ++$temp;
                    }
                }

                $tasklevel = (int) $task['OutlineLevel'];
				if ($tasklevel) {
					for($i = 0; $i < $tasklevel; $i++) {
						$output .= '&nbsp;&nbsp;';
					}
					$output .= '<img src="' . w2PfindImage('corner-dots.gif') . '" border="0" />&nbsp;';
				}

				$output .= $task['Name'];

                if ($milestone) {
                    $output .= '<img src="' . w2PfindImage('icons/milestone.gif', $m) . '" border="0" />';
                }
//TODO: the formatting for the dates should be better
                $output .= '</td>
							<td class="center">'.$task['Start'].'</td>
							<td class="center">'.$task['Finish'].'</td>
							<td class="center">';

                //This is a bizarre function i've made to associate the resources with their tasks
                //If there's only one resource associate to a task, it skip the while. If there's more than one
                //the loop take care of it until the last resource
                //strange but it works, that's my moto.
                if (!empty($task['Resources'])) {
                    $x=0;
                    $y=strpos($task['Resources'],';');
                    while (!empty($y)) {
                        $recurso=substr($task['Resources'],$x,($y-$x));
						$output .= '<div name="userRelated">';
                        $output.= arraySelect($resources, 'tasks['.$k.'][resources][]', 'size="1" class="text"', $recurso);
						$output .= '&nbsp;';
						$output .= arraySelect($percent, 'tasks['.$k.'][resources_alloc][]', 'size="1" class="text"', 100) . '%';
						$output .= '</div>';
                        $x=$y+1;
                        $y=strpos($task['Resources'],';',$x);
                    }
                } else {
					$output .= '<div name="userRelated">';
					$output.= arraySelect($resources, 'tasks['.$k.'][resources][]', 'size="1" class="text"', $recurso);
					$output .= '&nbsp;';
					$output .= arraySelect($percent, 'tasks['.$k.'][resources_alloc][]', 'size="1" class="text"', 100) . '%';
					$output .= '</div>';
				}
                $output .= '</td></tr>';
            }
        }
        $output .= '</table></td></tr>';

        $output .= '</table>';
        return $output;
    }

    public function loadFile($AppUI) {
        $filename = $_FILES['upload_file']['tmp_name'];

        $file = fopen($filename, "r");
        $filedata = fread($file, $_FILES['upload_file']['size']);
        fclose($file);

        if (substr_count($filedata, '<tasks>') < 1) {
            return false;
        }
        $x = strpos($filedata, '<calendar>');
        $header = substr($filedata, 0, $x);
        $summaryNode = $this->stripper("<summary ","/>",$filedata);
        $taskNodes = $this->stripper("<tasks>","</tasks>",$filedata);
        $endNodes = "</proj></project>";

        if (substr_count($filedata, '<resources>') < 1) {
            echo "<b>".$this->AppUI->_("Failure")."</b> ".$this->AppUI->_("impinfo")."<BR>";
            $filedata=$header.$summaryNode.$taskNodes.$endNodes;
        } else {
            $userNodes=$this->stripper("<resources>","</resources>",$filedata);
            $filedata=$header.$summaryNode.$userNodes.$taskNodes.$endNodes;
        }
        /*
         * O resultado esperado Ã© esse:
         * <project>
         * 		<proj attributes....>
         * 			<summary ...../>
         * 			<resources>
         * 				<resource id="0"......./>
         * 				...
         * 				<resource id="X"......./>
         * 			</resources>
         * 			<tasks>
         * 				<task id="0"......./>
         * 				...
         * 				<task id="X"......./>
         * 			</tasks>
         * 		</proj>
         * </project>
         */

        $this->scrubbedData = $filedata;
        return true;
    }
    /* Extrai uma determinada tag xml de uma string que com
     * conteúdo de um arquivo
     * @param    string    $startTag Tag de inicio
     *          string    $endTag    Tag de final
     *             string    $data Escopo onde vai ser procurado a tag
    */
    private function stripper($startTag,$endTag,$data) {
        $x=strpos($data, $startTag);
        $y=strpos($data, $endTag,$x)+strlen($endTag);
        $data = substr($data, $x, ($y-$x));
        return $data;
    }
    private function dur($duration) {
        //Not a very good duration function, just take the number of days and multiplies it for 8
        $Offset = strpos($duration, 'd');
        $x = substr($duration, 0, $Offset);
        return ($x*8);
    }

	/*
	 * This nasty little function builds the WBS id structure for import.
	 */
    private function montar_wbs($current_level) {
		$current_level = (int) $current_level;

		if ($this->lastLevel == $current_level) {
			$this->depthArray['l-'.$current_level]++;
		}

		if ($current_level > $this->lastLevel) {
			$this->depthArray['l-'.$current_level]++;
		} elseif ($current_level < $this->lastLevel) {
			$this->depthArray['l-'.$current_level]++;
			for($i = $current_level+1; $i <= $this->maxLevel; $i++) {
				unset($this->depthArray['l-'.$i]);
			}
		}

		$this->maxLevel = ($current_level > $this->maxLevel) ? $current_level : $this->maxLevel;
		$this->lastLevel = (int) $current_level;
		$wbs = implode($this->depthArray, '.');

		return $wbs;
    }
}